scripts/ci/run-ci-tests.sh

```diff
@@ -288,7 +288,8 @@ ip net add test
 
 # Rootless tests
 # Check if cap_checkpoint_restore is supported and also if unshare -c is supported.
-if capsh --supports=cap_checkpoint_restore && unshare -c /bin/true; then
+# Do not run this test in a container.
+if capsh --supports=cap_checkpoint_restore && unshare -c /bin/true && [ ! -e /run/.containerenv ]; then
```

Comment:
I can't understand why we can run root tests but not non-root. Pls write more details in the comment and in the commit message.

Reply:
As far as I understand this mount related:

```
mnt: Path `/usr/lib64/libc.so.6' resolved to `./' mountpoint"
```

I don't know if and how to correctly fix it, but it seems using a mount namespace in a mount namespace does not work as expected.

The reproducer for me is. This works without a container:
```
setcap cap_checkpoint_restore,cap_sys_ptrace+eip criu/criu
sudo --user=#65534 --group=#65534 unshare -Ucfpm --mount-proc -- bash -c "./test/zdtm.py run -t zdtm/static/maps00 -k always  -f h --rootless && true"
```
But it does not work in a container. @Snorch can you say if this should work and how?

Reply:
> mnt: Path `/usr/lib64/libc.so.6' resolved to `./' mountpoint"

This just says that the mentioned file is on the root filesystem. This is not an error message. Sharing some error message would be nice.

I didn't manage to get anywhere with the above reproduce:
```
[root@turmoil criu]# sudo --user=#65534 --group=#65534 unshare -Ucfpm --mount-proc -- bash -c "./test/zdtm.py run -t zdtm/static/maps00 -k always  -f h --rootless && true"
userns is supported
=== Run 1/1 ================ zdtm/static/maps00
========================= Run zdtm/static/maps00 in h ==========================
Start test
./maps00 --pidfile=maps00.pid --outfile=maps00.out --filename=maps00.test
Traceback (most recent call last):
  File "/criu/test/zdtm.py", line 2847, in <module>
    fork_zdtm()
  File "/criu/test/zdtm.py", line 2838, in fork_zdtm
    do_run_test(tinfo[0], tinfo[1], tinfo[2], tinfo[3])
  File "/criu/test/zdtm.py", line 1932, in do_run_test
    cr(cr_api, t, opts)
  File "/criu/test/zdtm.py", line 1577, in cr
    cr_api.set_test(test)
  File "/criu/test/zdtm.py", line 1121, in set_test
    os.makedirs(self.__dump_path)
  File "<frozen os>", line 215, in makedirs
  File "<frozen os>", line 225, in makedirs
PermissionError: [Errno 13] Permission denied: 'dump/zdtm/static/maps00'
##################################### FAIL #####################################
```
Just strange file permission problems, no mount problems.

Reply:
If you look at the run-ci.sh script you see a couple of chmod 777. You need to run those. Just before the sudo command there is also a rm which you probably need to run.

Reply:
> I don't know if and how to correctly fix it, but it seems using a mount namespace in a mount namespace does not work as expected.

Strictly speaking, it is unacceptable to skip test without knowing exact reasons. In this case, the ability of C/R-ing in a container is one of target use-cases.

Reply:
> Sharing some error message would be nice.

The following error messages are from the logs in CI.
https://cirrus-ci.com/task/5101821836918784
```
b"(00.011456) mnt: Path `/usr/lib64/libc.so.6' resolved to `./' mountpoint"
b'(00.011458) Error (criu/proc_parse.c:379): Failed to resolve mapping 7fb21b404000 filename'
b"(00.011561) Error (criu/proc_parse.c:693): Can't open 66's mapfile link 7fb21b404000: Operation not permitted"
b'(00.011570) Error (criu/cr-dump.c:1563): Collect mappings (pid: 66) failed with -1'
```

@adrianreber Looking at the CI logs below, is it possible that the test fails because selinux is enabled on the host and we can't disable it from within the container?
```
++ getenforce
+ selinuxmode=Disabled
+ setenforce Permissive
setenforce: SELinux is disabled
```
https://cirrus-ci.com/task/4965372655304704?logs=build#L12622

Reply:
> Strictly speaking, it is unacceptable to skip test without knowing exact reasons. In this case, the ability of C/R-ing in a container is one of target use-cases.

You are right. But my goal with this PR is to get CI green again. I think it is more important to have useful CI results than keeping broken tests running in CI. If a test is always red it might hide new errors introduced by new commits and becomes less useful. This test run alone had three broken tests and the first broken test was hiding the two follow-up errors. So this might have been merged without knowing that it does not work in all setups.

Anyway, if you prefer to not have this PR, no problem, I can close it.

Reply:
> @adrianreber Looking at the CI logs below, is it possible that the test fails because selinux is enabled on the host and we can't disable it from within the container?

SELinux does not change anything on my SELinux enabled system when I reproduce it locally. Initially I thought it might be related to SELinux. Outside of a container it is, SELinux blocks checkpoint/restore in a user namespace. But creating a user namespace inside of a container SELinux does not seem to affect the results. Do you see other results?

Reply:
I was able to replicate the error locally and does not seem to be related to SELinux.
It occurs only when running the test in user namespace and CRIU fails in `phys_stat_dev_match()` because the value returned from `phys_stat_resolve_dev()` is not equal to `phys_dev`.

https://github.com/checkpoint-restore/criu/blob/b17a73b2ee21d64a7cc59968f39e784618837397/criu/mount.c#L297

The test seems to work without user namespace:
```
$ unshare -fpm --mount-proc -- bash -c "./test/zdtm.py run -t zdtm/static/maps00 -f h --rootless && true"
userns is supported
=== Run 1/1 ================ zdtm/static/maps00
========================= Run zdtm/static/maps00 in h ==========================
Start test
./maps00 --pidfile=maps00.pid --outfile=maps00.out --filename=maps00.test
Run criu dump
=[log]=> dump/zdtm/static/maps00/66/1/dump.log
------------------------ grep Error ------------------------
b'(00.000000) CRIU run id = 0xf000058100000043'
b'(00.000141) Version: 3.18 (gitid 0)'
b'(00.000169) Running on e07aabb71f84 Linux 6.5.10-200.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Nov  2 19:59:55 UTC 2023 x86_64'
b'(00.000180) Warn  (criu/kerndat.c:1153): $XDG_RUNTIME_DIR not set. Cannot find location for kerndat file'
b'(00.022561) cpu: fpu: MPX CSR                          xstate_offsets   1024 / 832    xstate_sizes     64 / 64'
b'(00.023064) Effective capability 12 missing'
b'(00.024829) Found mmap_min_addr 0x10000'
b'(00.024880) files stat: fs/nr_open 1073741816'
b'(00.024893) Warn  (criu/kerndat.c:1153): $XDG_RUNTIME_DIR not set. Cannot find location for kerndat file'
------------------------ ERROR OVER ------------------------
Run criu restore
=[log]=> dump/zdtm/static/maps00/66/1/restore.log
------------------------ grep Error ------------------------
b'(00.000000) CRIU run id = 0xf00005810000004b'
b'(00.000109) Version: 3.18 (gitid 0)'
b'(00.000131) Running on e07aabb71f84 Linux 6.5.10-200.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Nov  2 19:59:55 UTC 2023 x86_64'
b'(00.000141) Warn  (criu/kerndat.c:1153): $XDG_RUNTIME_DIR not set. Cannot find location for kerndat file'
b'(00.021152) cpu: fpu: MPX CSR                          xstate_offsets   1024 / 832    xstate_sizes     64 / 64'
b'(00.021723) Effective capability 12 missing'
b'(00.023582) Found mmap_min_addr 0x10000'
b'(00.023641) files stat: fs/nr_open 1073741816'
b'(00.023655) Warn  (criu/kerndat.c:1153): $XDG_RUNTIME_DIR not set. Cannot find location for kerndat file'
------------------------ ERROR OVER ------------------------
Send the 15 signal to  66
Wait for zdtm/static/maps00(66) to die for 0.100000
Removing dump/zdtm/static/maps00/66
========================= Test zdtm/static/maps00 PASS =========================
```

Reply:
> It occurs only when running the test in user namespace and CRIU fails in phys_stat_dev_match() because the value returned from phys_stat_resolve_dev() is not equal to phys_dev.

It is one of the known overlayfs issues. Cc: @mihalicyn

Reply:
These changes have to fix the issue in the kernel:
https://github.com/avagin/linux-task-diag/commits/overlay-vs-maps

Reply:
See #2312. I will merge this with a comment mentioning the ticket and that it can be removed in the future.